### PR TITLE
[Lifecycle] [Lint] fixes b/184830263 Lint should use check reference too

### PR DESCRIPTION
**Pull request by kozaxinan**
## Proposed Changes

  - Fix NonNullableMutableLiveDataDetector for  [different cases](https://issuetracker.google.com/issues/184830263). It is possible to only set type reference and omit call's generic argument. 
  -  Added check for generic values because default upper bound of generic (if none specified) is Any?.  If generic is not specified as `<T : Any>`, `T` can be defined null at call site. https://kotlinlang.org/docs/generics.html#upper-bounds 
  

## Testing 

Test: Run NonNullableMutableLiveDataDetectorTest with new edge cases.
`nullLiteralFailMultipleFields` : Added type reference
`justKotlinObject` : Add additional test to prevent ArrayIndexOutOfBoundsException. It is fixed in https://github.com/androidx/androidx/commit/428f0ec685ff80035c0511c0cdb12b9770b6159c but not tested. It is reported here with [b/184830262](https://issuetracker.google.com/issues/184830262)
`genericParameterDefinition` : Generics are assumed as nullable, lint should ignore.

## Issues Fixed

Fixes: The bug on [b/184830263](https://issuetracker.google.com/issues/184830263) being fixed
Fixes: Tests [b/184830262](https://issuetracker.google.com/issues/184830263)

### CODE REVIEW

**Comment by kozaxinan** [comment]
@googlebot I signed it!

**Comment by kozaxinan** [comment]
@jbw0033 Is there anything I can add to this PR? 
This lint blocks us to update latest version of lifecycle.

File: lifecycle/lifecycle-livedata-core-ktx-lint/src/main/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetector.kt [hunk 1]
```diff
@@ -170,6 +183,9 @@ class NonNullableMutableLiveDataDetector : Detector(), UastScanner {
         context: JavaContext,
         node: UCallExpression
     ) {
+        // ignore generic generic MutableLiveData<T>
```
Comment (jbw0033) [inline lifecycle/lifecycle-livedata-core-ktx-lint/src/main/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetector.kt]: // ignore generic types

File: lifecycle/lifecycle-livedata-core-ktx-lint/src/test/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetectorTest.kt [hunk 2]
```diff
@@ -760,4 +765,40 @@ Fix for src/com/example/MyClass2.kt line 30: Change `LiveData` type to nullable:
             ).indented()
         ).expectClean()
     }
+
+    // https://issuetracker.google.com/issues/184830262
```
Comment (jbw0033) [inline lifecycle/lifecycle-livedata-core-ktx-lint/src/test/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetectorTest.kt]: You can remove the comment and just add that bug as another `Fixes: ` line for the PR.

File: lifecycle/lifecycle-livedata-core-ktx-lint/src/test/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetectorTest.kt [hunk 3]
```diff
@@ -760,4 +765,40 @@ Fix for src/com/example/MyClass2.kt line 30: Change `LiveData` type to nullable:
...
+                package com.example
+
+                import androidx.lifecycle.MutableLiveData
+
+                class Foo<T>(
+                    var target: MutableLiveData<T>
+                ) {
+                
+                    fun foo(value: T) {
+                        target.value = value
```
Comment (jbw0033) [inline lifecycle/lifecycle-livedata-core-ktx-lint/src/test/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetectorTest.kt]: Should this not be testing `target.value = null`?

File: lifecycle/lifecycle-livedata-core-ktx-lint/src/main/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetector.kt [hunk 4]
```diff
@@ -94,15 +96,26 @@ class NonNullableMutableLiveDataDetector : Detector(), UastScanner {
...
+                if (typeArgument != null) {
+                    return typeArgument
+                }
+
                 // We need to extract type from the expression
                 // Given the field `val liveDataField = MutableLiveData<Boolean>()`
                 // expression: `MutableLiveData<Boolean>()`
                 // argument: `Boolean`
-                val expression = element.sourcePsi
+                val callExpression = element.sourcePsi
```
Comment (jbw0033) [inline lifecycle/lifecycle-livedata-core-ktx-lint/src/main/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetector.kt]: keep this val name as `expression` or change the comment above. Either works.

**Review by jbw0033 — CHANGES_REQUESTED** [review]

File: lifecycle/lifecycle-livedata-core-ktx-lint/src/test/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetectorTest.kt (hunk 3 above)
Comment (kozaxinan) [reply lifecycle/lifecycle-livedata-core-ktx-lint/src/test/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetectorTest.kt]: value T is already nullable by default because <T> is defined without any type information. https://kotlinlang.org/docs/generics.html#upper-bounds

File: lifecycle/lifecycle-livedata-core-ktx-lint/src/main/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetector.kt (hunk 4 above)
Comment (kozaxinan) [reply lifecycle/lifecycle-livedata-core-ktx-lint/src/main/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetector.kt]: Reverted name change.

File: lifecycle/lifecycle-livedata-core-ktx-lint/src/test/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetectorTest.kt (hunk 2 above)
Comment (kozaxinan) [reply lifecycle/lifecycle-livedata-core-ktx-lint/src/test/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetectorTest.kt]: Description is updated and comment is removed.

File: lifecycle/lifecycle-livedata-core-ktx-lint/src/test/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetectorTest.kt (hunk 3 above)
Comment (jbw0033) [reply lifecycle/lifecycle-livedata-core-ktx-lint/src/test/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetectorTest.kt]: Right, but we already know a value with type T will match just fine, what we actually want to verify here is that the lint rule will ignore setting a variable with a type of T to null since T is indeed nullable.

**Review by jbw0033 — CHANGES_REQUESTED** [review]

File: lifecycle/lifecycle-livedata-core-ktx-lint/src/test/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetectorTest.kt (hunk 3 above)
Comment (kozaxinan) [reply lifecycle/lifecycle-livedata-core-ktx-lint/src/test/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetectorTest.kt]: Test is updated.

**Review by jbw0033 — APPROVED** [review]

**Comment by jbw0033** [comment]
From our internal lint:

lifecycle/lifecycle-livedata-core-ktx-lint/src/test/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetectorTest.kt
#202
Patchset 1
9:08 AM
Lint 🤖
Please remove the trailing whitespace.

#794
Patchset 1
9:08 AM
Lint 🤖
Please remove the trailing whitespace.

**Comment by kozaxinan** [comment]
> From our internal lint:
> 
> lifecycle/lifecycle-livedata-core-ktx-lint/src/test/java/androidx/lifecycle/lint/NonNullableMutableLiveDataDetectorTest.kt
> #202
> Patchset 1
> 9:08 AM
> Lint 🤖
> Please remove the trailing whitespace.
> 
> #794
> Patchset 1
> 9:08 AM
> Lint 🤖
> Please remove the trailing whitespace.

I deleted whitespaces. When I run `./gradlew buildOnServer` some unrelated lints fails, is that expected? (even on main branch)

**Comment by jbw0033** [comment]
Thank you for the pull request!
